numcodecs/vlen.pyx

```diff
@@ -0,0 +1,443 @@
+        encoded_values = np.empty(n_items, dtype=object)
+        encoded_lengths = np.empty(n_items, dtype=np.intc)
+
+        # first iteration to convert to bytes
+        data_length = 0
+        for i in range(n_items):
+            u = input_values[i]
+            if not PyUnicode_Check(u):
+                raise TypeError('expected unicode string, found %r' % u)
+            b = PyUnicode_AsUTF8String(u)
+            l = PyBytes_GET_SIZE(b)
+            encoded_values[i] = b
+            data_length += l + 4  # 4 bytes to store item length
+            encoded_lengths[i] = l
+
+        # setup output
+        total_length = HEADER_LENGTH + data_length
+        out = PyByteArray_FromStringAndSize(NULL, total_length)
+
+        # write header
+        data = PyByteArray_AS_STRING(out)
+        store_le32(data, n_items)
+
+        # second iteration, store data
+        data += HEADER_LENGTH
+        for i in range(n_items):
+            l = encoded_lengths[i]
+            store_le32(data, l)
+            data += 4
+            encv = PyBytes_AS_STRING(encoded_values[i])
+            memcpy(data, encv, l)
+            data += l
+
+        return out
+
+    @cython.wraparound(False)
+    @cython.boundscheck(False)
+    def decode(self, buf, out=None):
+        cdef:
+            Buffer input_buffer
+            char* data
+            char* data_end
+            Py_ssize_t i, l, n_items, data_length, input_length
+
+        # accept any buffer
+        input_buffer = Buffer(buf, PyBUF_ANY_CONTIGUOUS)
+        input_length = input_buffer.nbytes
+
+        # sanity checks
+        if input_length < HEADER_LENGTH:
+            raise ValueError('corrupt buffer, missing or truncated header')
+
+        # obtain input data pointer
+        data = input_buffer.ptr
+        data_end = data + input_length
+
+        # load number of items
+        n_items = load_le32(data)
+
+        # setup output
+        if out is not None:
+            out = check_out_param(out, n_items)
+        else:
+            out = np.empty(n_items, dtype=object)
+
+        # iterate and decode - N.B., do not try to cast `out` as object[:]
+        # as this causes segfaults, possibly similar to
+        # https://github.com/cython/cython/issues/1608
+        data += HEADER_LENGTH
+        for i in range(n_items):
+            if data + 4 > data_end:
+                raise ValueError('corrupt buffer, data seem truncated')
+            l = load_le32(data)
+            data += 4
+            if data + l > data_end:
+                raise ValueError('corrupt buffer, data seem truncated')
+            out[i] = PyUnicode_FromStringAndSize(data, l)
+            data += l
+
+        return out
+
+
+class VLenBytes(Codec):
+    """Encode variable-length byte string objects.
+
+    Examples
+    --------
+    >>> import numcodecs
+    >>> import numpy as np
+    >>> x = np.array([b'foo', b'bar', b'baz'], dtype='object')
+    >>> codec = numcodecs.VLenUTF8()
```

Comment:
Just caught this mismatch when reading the docs, added PR ( https://github.com/alimanfoo/numcodecs/pull/58 ) to fix it.